sklearn/cluster/_kmeans.py

```diff
@@ -395,14 +396,24 @@ def _kmeans_single_elkan(X, sample_weight, centers_init, max_iter=300,
 
         centers, centers_new = centers_new, centers
 
-        center_shift_tot = (center_shift**2).sum()
-        if center_shift_tot <= tol:
-            if verbose:
-                print(f"Converged at iteration {i}: center shift "
-                      f"{center_shift_tot} within tolerance {tol}.")
-            break
+        if tol == 0:
+            # When tol = 0 we check that labels did not change because
+            # center_shift might not be exactly 0 due to rounding errors.
+            if np.array_equal(labels, labels_old):
+                strict_convergence = True
+                break
+            labels_old = labels.copy()
+        else:
+            center_shift_tot = (center_shift**2).sum()
+            if center_shift_tot <= tol:
+                if verbose:
+                    print(f"Converged at iteration {i}: center shift "
+                          f"{center_shift_tot} within tolerance {tol}.")
+                strict_convergence = center_shift_tot == 0
+                break
+        strict_convergence = False
```

Comment:
If `center_shift_tot <= tol` at the first iteration, then `strict_convergence` will be undefined. I think I would find it more natural to define:

```python
strict_convergence = False
```
before entering the for `for i in range(max_iter)`.

Reply:
It will be defined since I define strict_convergence before the break.
And max_iter is > 0 so we are sure to enter the loop.

But I'm ok to define it before the loop if you find it clearer.

Reply:
Thanks.